tree/forest/v7/inc/ROOT/RTree.hxx

```diff
@@ -0,0 +1,148 @@
+
+// clang-format off
+/**
+\class ROOT::Experimental::RTree
+\ingroup Forest
+\brief The RTree represents a live dataset, whose structure is defined by an RTreeModel
+
+RTree connects the static information of the RTreeModel to a source or sink on physical storage.
+Reading and writing requires use of the corresponding derived class RInputTree or ROutputTree.
+RTree writes only complete entries (rows of the data set).  The entry itself is not kept within the
+RTree, which allows for multiple concurrent entries for the same RTree.  Besides reading an entire entry,
+the RTree can expose tree views that read only specific branches.
+*/
+// clang-format on
+class RTree {
+private:
+   /// All trees that use the same model share its ownership
+   std::shared_ptr<RTreeModel> fModel;
+   /// The number of entries is constant for reading and reflects the current number of Fill() operations
+   /// when writing
+   TreeIndex_t fNentries;
+
+protected:
+   /// Only the derived RInputTree and ROutputTree can be instantiated
+   explicit RTree(std::shared_ptr<RTreeModel> model);
+
+public:
+   RTree(const RTree&) = delete;
+   RTree& operator =(const RTree&) = delete;
+   ~RTree();
+}; // RTree
+
+} // namespace Detail
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::RInputTree
+\ingroup Forest
+\brief An RTree that is used to read data from storage
+
+An input tree provides data from storage as C++ objects. The tree model can be created from the data on storage
+or it can be imposed by the user. The latter case allows users to read into a specialized tree model that covers
+only a subset of the branches in the tree. The tree model is used when reading complete entries.
+Individual branches can be read as well by instantiating a tree view.
+*/
+// clang-format on
+class RInputTree : public Detail::RTree {
+private:
+   std::unique_ptr<Detail::RTreeSource> fSource;
+
+public:
+   /// The user imposes a tree model, which must be compatible with the model found in the data on storage
+   RInputTree(std::shared_ptr<RTreeModel> model, std::unique_ptr<Detail::RTreeSource> source);
+   /// The model is generated from the tree metadata on storage
+   RInputTree(std::unique_ptr<Detail::RTreeSource> source);
+   ~RInputTree();
+
+   /// Analogous to Fill(), fills the default entry of the model
+   void GetEntry();
+   /// Fills a user provided entry after checking that the entry has been instantiated from the tree's model
+   void GetEntry(RTreeEntry &entry);
+
+   /// Provides access to an individual branch that can be either a leaf or a collection, e.g.
+   /// GetView<double>("particles.pt") or GetView<RVec<double>>("particle").  It can as well be the offset
+   /// branch of a collection itself, like GetView<TreeIndex_t>("particle")
+   template <typename T>
+   RTreeView<T> GetView(std::string_view branchName) {
+      auto branch = std::make_unique<RBranch<T>>(branchName, fSource.get());
+      // ...
+      return RTreeView<T>(std::move(branch));
```

Comment:
Looks like `fSource.get()` is escaping the memory management with a raw pointer. Is the source guaranteed to not vanish before `RBranch<T>` is destroyed? Another thing is that, looking at the code, `RBranch` takes a reference to `RTreeSource`, and the thing being passed here is a pointer, which can be `nullptr`. I don't see this being accounted for in the code.